File: android/AgentShell/src/main/java/com/agentos/shell/screens/RichOutput.kt
package com.agentos.shell.screens

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.agentos.shell.theme.T

/**
 * Rich Visual Outputs — turns a plain answer into an elegant hero card when it leads with a headline
 * value worth seeing first: weather, a price, a stock move, or a game score. Design language is quiet and
 * premium (soft gradient surface, hairline border, generous space, one big number) rather than a loud
 * filled box. Detection stays conservative so ordinary answers are never restyled.
 */
sealed class Hero {
    data class Metric(val eyebrow: String, val big: String, val unit: String, val sub: String) : Hero()
    data class Stock(val eyebrow: String, val price: String, val delta: String, val up: Boolean, val sub: String) : Hero()
    data class Score(val teamA: String, val a: Int, val teamB: String, val b: Int) : Hero()
    data class YesNo(val yes: Boolean, val sub: String) : Hero()
    data class Quote(val text: String, val author: String) : Hero()
}

private val GREEN = Color(0xFF1FA855)

object RichParse {
    private val temp = Regex("(-?\\d{1,3})\\s*°\\s*([CFcf])")
    private val price = Regex("\\$\\s?([0-9][0-9,]{0,12}(?:\\.[0-9]{1,2})?)")
    private val pctSigned = Regex("([+-]\\d{1,3}(?:\\.\\d+)?)\\s*%")
    private val pctWord = Regex("(?i)\\b(up|down|gained|lost|fell|rose|climbed|dropped|jumped|slid|surged)\\b[^%\\n]{0,14}?(\\d{1,3}(?:\\.\\d+)?)\\s*%")
    private val score = Regex("([A-Z][A-Za-z.&'0-9 ]{1,20}?)\\s+(\\d{1,3})\\s*[-–—]\\s*(\\d{1,3})\\s+([A-Z][A-Za-z.&'0-9 ]{1,20})")
    private val gameWord = Regex("(?i)\\b(beat|defeat|final|score|won|win|vs\\.?|versus|match|game|quarter|half[- ]?time|inning|goals?|touchdown|nba|nfl|mlb|nhl|premier league|la liga)\\b")
    private val stopTickers = setOf("THE", "AND", "FOR", "USD", "CEO", "ETF", "IPO", "USA", "GDP", "API", "AI", "PM", "AM", "EPS", "YOY", "Q1", "Q2", "Q3", "Q4")
    // "3 days until launch", "2 weeks left" — a clear time-to-event headline.
    private val countdown = Regex("(?i)\\b(\\d{1,4})\\s*(day|days|week|weeks|hour|hours|month|months)\\s+(?:until|till|til|to go|left|remaining|away|before)\\b")
    // A rating like "4.5 out of 5" or "9/10" — only trusted when the text is clearly about a rating.
    private val ratingCtx = Regex("(?i)\\b(star|stars|rating|rated|review|reviews)\\b")
    private val rating = Regex("\\b(\\d(?:\\.\\d)?)\\s*(?:/|out of)\\s*(\\d{1,2})\\b")
    // Currency conversion — "€100 = $108", "£50 ≈ $63".
    private val convert = Regex("([$€£¥]\\s?[0-9][0-9,]*(?:\\.[0-9]+)?)\\s*(?:=|≈|~|is about|is roughly|equals?)\\s*([$€£¥]\\s?[0-9][0-9,]*(?:\\.[0-9]+)?)")
    // Time in a place — "3:45 PM in Tokyo".
    private val timeIn = Regex("(?i)\\b(\\d{1,2}:\\d{2}\\s*(?:[ap]\\.?m\\.?)?)\\s+in\\s+([A-Z][A-Za-z .'-]{2,28})")
    // A bare calculation result — a short reply that ends in "= number".
    private val mathEq = Regex("=\\s*(-?[0-9][0-9,]*(?:\\.[0-9]+)?)\\s*\\.?\\s*$")
    // Translation — "in Spanish: hola" / "in French is « bonjour »".
    private val translate = Regex("(?i)\\bin (spanish|french|german|italian|portuguese|japanese|chinese|korean|russian|arabic|hindi|dutch|greek|latin)\\b\\s*(?:is|:|,)?\\s*[\"“«']?([\\p{L} ]{1,40})[\"”»']?")
    // A quotation with attribution — "…" — Author.
    private val quote = Regex("[\"“](.{12,220}?)[\"”]\\s*[—–-]\\s*([A-Z][A-Za-z. '-]{2,40})")
    // Distance / weight with a unit, near the start.
    private val measure = Regex("(?i)\\b([0-9][0-9,]*(?:\\.[0-9]+)?)\\s*(km|kilometers?|miles?|kg|kilograms?|pounds?|lbs)\\b")

    /** A hero card for [reply], or null to fall back to plain text. */
    fun detect(reply: String): Hero? {
        val r = reply.trim()
        if (r.length < 3) return null

        temp.find(r)?.takeIf { it.range.first < 60 }?.let { m ->
            val (t, u) = m.destructured
            return Hero.Metric("Weather", "$t°${u.uppercase()}", "", firstSentence(r))
        }

        // Countdown — "3 days until launch", "2 weeks left".
        countdown.find(r)?.takeIf { it.range.first < 90 }?.let { m ->
            val (num, unit) = m.destructured
            return Hero.Metric("Countdown", num, unit.lowercase(), firstSentence(r))
        }

        // Rating — only when the text clearly talks about a rating/review (avoids "5/10 people").
        if (ratingCtx.containsMatchIn(r)) rating.find(r)?.let { m ->
            val (a, b) = m.destructured
            return Hero.Metric("Rating", "$a/$b", "", firstSentence(r))
        }

        // Quotation with attribution — "…" — Author.
        quote.find(r)?.let { m ->
            val (txt, author) = m.destructured
            return Hero.Quote(txt.trim(), author.trim())
        }

        // Currency conversion — show the converted value big.
        convert.find(r)?.takeIf { it.range.first < 90 }?.let { m ->
            val (from, to) = m.destructured
            return Hero.Metric("Conversion", to.trim(), "", from.trim() + " = " + to.trim())
        }

        // Time in a place — "3:45 PM in Tokyo".
        timeIn.find(r)?.takeIf { it.range.first < 60 }?.let { m ->
            val (t, place) = m.destructured
            return Hero.Metric("Time · " + place.trim(), t.trim().uppercase(), "", firstSentence(r))
        }

        // Translation — a single word/phrase in another language.
        translate.find(r)?.let { m ->
            val lang = m.groupValues[1].replaceFirstChar { it.uppercase() }
            val word = m.groupValues[2].trim()
            if (word.length in 1..40) return Hero.Metric(lang, word, "", firstSentence(r))
        }

        // Measurement — distance / weight with a unit near the start.
        measure.find(r)?.takeIf { it.range.first < 55 }?.let { m ->
            val (n, unit) = m.destructured
            return Hero.Metric("Measurement", n, unit.lowercase(), firstSentence(r))
        }

        // Bare calculation result — only for a short reply ending in "= number".
        if (r.length < 70) mathEq.find(r)?.let { m ->
            return Hero.Metric("Result", m.groupValues[1], "", firstSentence(r))
        }

        // Game score — only when the text clearly reads like a match result (avoids "iPhone 15 - 128 GB").
        if (gameWord.containsMatchIn(r)) score.find(r)?.takeIf { it.range.first < 110 }?.let { m ->
            val (ta, sa, sb, tb) = m.destructured
            val ai = sa.toIntOrNull(); val bi = sb.toIntOrNull()
            if (ai != null && bi != null) return Hero.Score(ta.trim(), ai, tb.trim(), bi)
        }

        // Price present → either a stock move (price + %) or a plain price headline.
        price.find(r)?.takeIf { it.range.first < 60 }?.let { m ->
            val priceStr = "$" + m.groupValues[1]
            val signed = pctSigned.find(r)
            val word = pctWord.find(r)
            if (signed != null || word != null) {
                val (delta, up) = when {
                    signed != null -> signed.groupValues[1] to !signed.groupValues[1].startsWith("-")
                    else -> {
                        val w = word!!.groupValues[1].lowercase()
                        val n = word.groupValues[2]
                        val u = w in setOf("up", "gained", "rose", "climbed", "jumped", "surged")
                        (if (u) "+$n" else "-$n") to u
                    }
                }
                return Hero.Stock(tickerOf(r), priceStr, "$delta%", up, firstSentence(r))
            }
            return Hero.Metric("Price", priceStr, "", firstSentence(r))
        }

        // Yes / No — a clear affirmative or negative opener on a concise answer.
        //
        // A giant YES or NO is the most assertive thing this app can put on screen, so it has to be
        // right. Two ways it was wrong:
        //
        //  1. NUANCE. "No, he's not in your contacts, but I did find him in your messages" opened with
        //     "No" and got a huge NO headline sitting directly above a body saying the person WAS
        //     found. The card flatly contradicted the answer underneath it. Any pivot word means the
        //     answer is qualified, and a qualified answer has no business being reduced to one word.
        //  2. IDIOM. "No problem", "No worries", "Yes please" aren't answers to a yes/no question at
        //     all — they're conversational filler that happens to start with the right token.
        if (r.length < 280) Regex("(?i)^(yes|yep|yeah|correct|absolutely|no|nope|nah)\\b[,.!: ]").find(r)?.let { m ->
            val w = m.groupValues[1].lowercase()
            val idiom = Regex("(?i)^(no (problem|worries|need|rush|idea)|yes please|yeah sure)\\b").containsMatchIn(r)
            val qualified = Regex("(?i)\\b(but|however|although|though|that said|on the other hand|technically|" +
                "sort of|kind of|depends|unless|except)\\b").containsMatchIn(r)
            if (!idiom && !qualified) {
                return Hero.YesNo(w in setOf("yes", "yep", "yeah", "correct", "absolutely"), firstSentence(r))
            }
        }
        return null
    }

    /**
     * The reliable path: the model may prefix an answer with a machine card tag it controls, e.g.
     * `[[card:score;Warriors;121;Lakers;118]]`. We parse that (100% reliable, phrasing-independent) and
     * strip it from the displayed text. Types: score, stat, stock, quote, yesno.
     */
    fun fromTag(reply: String): Pair<Hero?, String> {
        val m = Regex("^\\s*\\[\\[card:([^\\]]+)\\]\\]\\s*", RegexOption.IGNORE_CASE).find(reply) ?: return null to reply
        val body = reply.removeRange(m.range).trim()
        val p = m.groupValues[1].split(";").map { it.trim() }
        fun g(i: Int) = p.getOrElse(i) { "" }
        val hero: Hero? = when (g(0).lowercase()) {
            "score" -> { val a = g(2).toIntOrNull(); val b = g(4).toIntOrNull(); if (a != null && b != null) Hero.Score(g(1), a, g(3), b) else null }
            "stat", "metric" -> if (g(2).isNotBlank()) Hero.Metric(g(1).ifBlank { "" }, g(2), g(3), g(4)) else null
            "stock" -> if (g(2).isNotBlank()) Hero.Stock(g(1).ifBlank { "Market" }, g(2), g(3), !g(3).startsWith("-"), g(4)) else null
            "quote" -> if (g(1).isNotBlank()) Hero.Quote(g(1), g(2)) else null
            "yesno" -> Hero.YesNo(g(1).lowercase().startsWith("y"), g(2))
            else -> null
        }
        return hero to body
    }

    /** Card from the model's tag if present, otherwise from best-effort detection — plus the clean text. */
    fun render(reply: String): Pair<Hero?, String> {
        val (tagHero, body) = fromTag(reply)
        if (tagHero != null) return tagHero to body
        return detect(body) to body
    }

    private fun tickerOf(s: String): String {
        val m = Regex("\\b([A-Z]{2,5})\\b").findAll(s).map { it.value }.firstOrNull { it !in stopTickers }
        return m ?: "Market"
    }

    /**
     * The card's one-line subtitle.
     *
     * The old version cut at a hard character count, which lands mid-word — a subtitle about a venture
     * investor rendered as "VC S", which looks like a rendering fault rather than an abbreviation. If
     * there's no sentence break to use, back up to the last whole word and mark the cut with an
     * ellipsis so it reads as deliberately shortened.
     */
    private fun firstSentence(s: String): String {
        val cut = s.indexOfFirst { it == '.' || it == '\n' }
        if (cut in 12..96) return s.substring(0, cut).trim()
        val t = s.trim()
        if (t.length <= 84) return t
        val slice = t.substring(0, 84)
        val lastSpace = slice.lastIndexOf(' ')
        // Only honour the word boundary if it doesn't gut the line — otherwise a single long token
        // would leave us with almost nothing.
        val body = if (lastSpace >= 40) slice.substring(0, lastSpace) else slice
        return body.trimEnd().trimEnd(',', ';', ':', '-', '—') + "…"
    }
}

/** The elegant headline card shown above a Home answer when [RichParse] finds a hero value. */
@Composable
fun HeroCardView(hero: Hero) {
    val surface = Brush.verticalGradient(listOf(T.bgElevated, T.accentSoft.copy(alpha = 0.22f)))
    Column(
        Modifier.fillMaxWidth()
            .clip(RoundedCornerShape(22.dp))
            .background(surface)
            .border(1.dp, T.hairline, RoundedCornerShape(22.dp))
            .padding(horizontal = 24.dp, vertical = 22.dp)
    ) {
        when (hero) {
            is Hero.Metric -> {
                Eyebrow(hero.eyebrow)
                Spacer(Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.Bottom) {
                    Text(hero.big, fontSize = 52.sp, fontWeight = FontWeight.Light, color = T.ink, letterSpacing = (-1).sp)
                    if (hero.unit.isNotBlank()) {
                        Spacer(Modifier.width(4.dp))
                        Text(hero.unit, fontSize = T.body, color = T.inkSoft, modifier = Modifier.padding(bottom = 10.dp))
                    }
                }
                if (hero.sub.isNotBlank() && !hero.sub.equals(hero.eyebrow, true)) {
                    Spacer(Modifier.height(8.dp)); Text(hero.sub, fontSize = T.small, color = T.inkSoft)
                }
            }
            is Hero.Stock -> {
                Eyebrow(hero.eyebrow)
                Spacer(Modifier.height(10.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Text(hero.price, fontSize = 46.sp, fontWeight = FontWeight.Light, color = T.ink, letterSpacing = (-1).sp)
                    Spacer(Modifier.width(14.dp))
                    DeltaPill(hero.delta, hero.up)
                }
                if (hero.sub.isNotBlank()) { Spacer(Modifier.height(8.dp)); Text(hero.sub, fontSize = T.small, color = T.inkSoft) }
            }
            is Hero.Score -> {
                Eyebrow("Final score")
                Spacer(Modifier.height(14.dp))
                TeamRow(hero.teamA, hero.a, hero.a >= hero.b)
                Spacer(Modifier.height(10.dp))
                TeamRow(hero.teamB, hero.b, hero.b >= hero.a)
            }
            is Hero.YesNo -> {
                val c = if (hero.yes) GREEN else T.danger
                Eyebrow("Answer")
                Spacer(Modifier.height(8.dp))
                Text(if (hero.yes) "Yes" else "No", fontSize = 48.sp, fontWeight = FontWeight.Bold, color = c, letterSpacing = (-1).sp)
                if (hero.sub.isNotBlank()) { Spacer(Modifier.height(8.dp)); Text(hero.sub, fontSize = T.small, color = T.inkSoft) }
            }
            is Hero.Quote -> {
                Row {
                    Box(Modifier.width(3.dp).heightIn(min = 30.dp).clip(RoundedCornerShape(999.dp)).background(T.accent))
                    Spacer(Modifier.width(14.dp))
                    Column {
                        Text("“" + hero.text + "”", fontSize = 20.sp, fontWeight = FontWeight.Light, color = T.ink, fontStyle = FontStyle.Italic)
                        Spacer(Modifier.height(8.dp))
                        Text("— " + hero.author, fontSize = T.small, fontWeight = FontWeight.Medium, color = T.accent)
                    }
                }
            }
        }
    }
}

@Composable
private fun Eyebrow(text: String) {
    Text(text.uppercase(), fontSize = 11.sp, fontWeight = FontWeight.SemiBold, color = T.accent, letterSpacing = 2.sp)
}

@Composable
private fun DeltaPill(delta: String, up: Boolean) {
    val c = if (up) GREEN else T.danger
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.clip(RoundedCornerShape(999.dp)).background(c.copy(alpha = 0.14f)).padding(horizontal = 12.dp, vertical = 6.dp)
    ) {
        Text(if (up) "▲" else "▼", fontSize = 11.sp, color = c)
        Spacer(Modifier.width(5.dp))
        Text(delta, fontSize = T.small, fontWeight = FontWeight.SemiBold, color = c)
    }
}

@Composable
private fun TeamRow(team: String, score: Int, winner: Boolean) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.SpaceBetween) {
        Row(verticalAlignment = Alignment.CenterVertically, modifier = Modifier.weight(1f)) {
            if (winner) { Box(Modifier.width(3.dp).height(22.dp).clip(RoundedCornerShape(999.dp)).background(T.accent)); Spacer(Modifier.width(10.dp)) }
            else Spacer(Modifier.width(13.dp))
            Text(team, fontSize = T.body, fontWeight = if (winner) FontWeight.SemiBold else FontWeight.Normal,
                color = if (winner) T.ink else T.inkSoft)
        }
        Text("$score", fontSize = 30.sp, fontWeight = if (winner) FontWeight.SemiBold else FontWeight.Light,
            color = if (winner) T.ink else T.inkSoft)
    }
}

File: ios/SlyOS/Sources/SlyOSCompanion/RichOutput.swift
import SwiftUI

/// Turns a plain answer into something worth looking at — a port of Compose `RichOutput`.
///
/// When an answer has a shape SlyOS recognises — a countdown, a price, a conversion, a quotation, a
/// straight yes or no — it gets a headline card above the prose. Everything else renders as prose,
/// because a card around an ordinary paragraph is decoration, not design.
enum Hero: Equatable {

    /// One big number with a label and optional unit: countdowns, conversions, measurements, times.
    case metric(eyebrow: String, big: String, unit: String, sub: String)
    /// A price with a movement.
    case stock(eyebrow: String, price: String, delta: String, up: Bool, sub: String)
    /// A straight answer to a straight question.
    case yesNo(yes: Bool, sub: String)
    /// A quotation with attribution.
    case quote(text: String, author: String)

    /// Find a headline in an answer, if there honestly is one.
    ///
    /// Every rule here is deliberately narrow. A wrong card is far worse than no card: it is the
    /// most assertive thing the app can put on screen, and it sits directly above prose that may
    /// contradict it.
    static func detect(in raw: String) -> Hero? {
        let text = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, text.count < 900 else { return nil }
        let firstLine = text.split(separator: "\n").first.map(String.init) ?? text

        // Quotation with attribution — “…” — Author
        if let m = firstLine.firstMatch(#"^[“"](.{10,220})[”"]\s*[—–-]\s*(.{2,60})$"#) {
            return .quote(text: m[1], author: m[2].trimmingCharacters(in: .whitespaces))
        }

        // Countdown — "3 days until launch", "2 weeks left"
        if let m = firstLine.firstMatch(#"(?i)\b(\d+)\s+(second|minute|hour|day|week|month|year)s?\b\s*(until|till|to go|left|remaining)"#) {
            return .metric(eyebrow: "Countdown", big: m[1], unit: m[2] + (m[1] == "1" ? "" : "s"),
                           sub: firstLine)
        }

        // Currency conversion — "€100 = $108"
        if let m = firstLine.firstMatch(#"([€£$¥]\s?[\d,.]+)\s*(?:=|≈|is)\s*([€£$¥]\s?[\d,.]+)"#) {
            return .metric(eyebrow: m[1], big: m[2], unit: "", sub: firstLine)
        }

        // A price with a movement — stock-shaped answers.
        if let price = firstLine.firstMatch(#"([€£$¥]\s?[\d,]+(?:\.\d+)?)"#),
           let delta = firstLine.firstMatch(#"([+-]\s?\d+(?:\.\d+)?\s?%)"#) {
            let up = !delta[1].contains("-")
            return .stock(eyebrow: subject(of: firstLine), price: price[1],
                          delta: delta[1].replacingOccurrences(of: " ", with: ""), up: up, sub: firstLine)
        }

        // Time in a place — "3:45 PM in Tokyo"
        if let m = firstLine.firstMatch(#"(?i)\b(\d{1,2}:\d{2}\s?(?:AM|PM)?)\b.*?\bin\s+([A-Z][\w\s]{2,24})"#) {
            return .metric(eyebrow: "Time in " + m[2].trimmingCharacters(in: .whitespaces),
                           big: m[1], unit: "", sub: "")
        }

        // A bare calculation — a short reply ending in "= number".
        if firstLine.count < 60, let m = firstLine.firstMatch(#"=\s*([\d,]+(?:\.\d+)?)\s*$"#) {
            return .metric(eyebrow: "Result", big: m[1], unit: "", sub: firstLine)
        }

        if let yesNo = detectYesNo(text, firstLine: firstLine) { return yesNo }
        return nil
    }

    /// Yes / No, with the two guards that stopped it being wrong.
    ///
    /// 1. **Nuance.** "No, he's not in your contacts, but I did find him in your messages" opens with
    ///    "No" and would get a giant NO sitting directly above a body saying the person *was* found.
    ///    Any pivot word means the answer is qualified, and a qualified answer has no business being
    ///    reduced to one word.
    /// 2. **Idiom.** "No problem", "No worries", "Yes please" are not answers to a yes/no question at
    ///    all — they are filler that happens to start with the right token.
    private static func detectYesNo(_ text: String, firstLine: String) -> Hero? {
        guard text.count < 320 else { return nil }
        let lower = firstLine.lowercased()

        for idiom in ["no problem", "no worries", "yes please", "no idea", "yes and no", "no longer",
                      "no need", "not sure", "no doubt"] where lower.hasPrefix(idiom) {
            return nil
        }
        for pivot in [" but ", " however", " although", " though ", " except", " unless ",
                      " technically", " sort of", " kind of", " depends"] where lower.contains(pivot) {
            return nil
        }

        guard let m = lower.firstMatch(#"^(yes|no|yep|nope|correct|incorrect)\b[,.!]?\s*(.*)$"#) else {
            return nil
        }
        let yes = ["yes", "yep", "correct"].contains(m[1])
        // The rest of the first line becomes the subtitle, so the card explains itself.
        let sub = String(firstLine.dropFirst(m[1].count)
            .trimmingCharacters(in: CharacterSet(charactersIn: " ,.!")))
        return .yesNo(yes: yes, sub: sub)
    }

    /// A rough label for what a price answer is about — whatever precedes the number.
    private static func subject(of line: String) -> String {
        let head = line.prefix { !$0.isNumber && $0 != "€" && $0 != "£" && $0 != "$" && $0 != "¥" }
        let cleaned = head.trimmingCharacters(in: CharacterSet(charactersIn: " ,:—-"))
        return cleaned.isEmpty ? "Price" : String(cleaned.prefix(40))
    }
}

/// The headline card.
struct HeroCardView: View {
    let hero: Hero
    @Environment(\.palette) private var p

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            switch hero {
            case let .metric(eyebrow, big, unit, sub):
                Eyebrow(eyebrow)
                HStack(alignment: .lastTextBaseline, spacing: 4) {
                    Text(big)
                        .font(.system(size: 52, weight: .light))
                        .tracking(-1)
                        .foregroundStyle(p.ink)
                    if !unit.isEmpty {
                        Text(unit).font(.system(size: T.body)).foregroundStyle(p.inkSoft)
                    }
                }
                .padding(.top, 10)
                if !sub.isEmpty && sub.lowercased() != eyebrow.lowercased() {
                    Text(sub).font(.system(size: T.small)).foregroundStyle(p.inkSoft).padding(.top, 8)
                }

            case let .stock(eyebrow, price, delta, up, sub):
                Eyebrow(eyebrow)
                HStack(spacing: 14) {
                    Text(price)
                        .font(.system(size: 46, weight: .light))
                        .tracking(-1)
                        .foregroundStyle(p.ink)
                    Text(delta)
                        .font(.system(size: T.small, weight: .medium))
                        .foregroundStyle(up ? p.good : p.danger)
                        .padding(.horizontal, 10).padding(.vertical, 4)
                        .background(Capsule().fill((up ? p.good : p.danger).opacity(0.15)))
                }
                .padding(.top, 10)
                if !sub.isEmpty {
                    Text(sub).font(.system(size: T.small)).foregroundStyle(p.inkSoft).padding(.top, 8)
                }

            case let .yesNo(yes, sub):
                Eyebrow("Answer")
                Text(yes ? "Yes" : "No")
                    .font(.system(size: 48, weight: .bold))
                    .tracking(-1)
                    .foregroundStyle(yes ? p.good : p.danger)
                    .padding(.top, 8)
                if !sub.isEmpty {
                    Text(sub).font(.system(size: T.small)).foregroundStyle(p.inkSoft).padding(.top, 8)
                }

            case let .quote(text, author):
                HStack(alignment: .top, spacing: 14) {
                    Capsule().fill(p.accent).frame(width: 3).frame(minHeight: 30)
                    VStack(alignment: .leading, spacing: 8) {
                        Text("“\(text)”")
                            .font(.system(size: 20, weight: .light))
                            .italic()
                            .foregroundStyle(p.ink)
                        Text("— \(author)")
                            .font(.system(size: T.small, weight: .medium))
                            .foregroundStyle(p.accent)
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 24).padding(.vertical, 22)
        .background(
            RoundedRectangle(cornerRadius: 22)
                .fill(LinearGradient(colors: [p.bgElevated, p.accentSoft.opacity(0.22)],
                                     startPoint: .top, endPoint: .bottom))
        )
        .overlay(RoundedRectangle(cornerRadius: 22).stroke(p.hairline, lineWidth: 1))
    }

    private func Eyebrow(_ s: String) -> some View {
        Text(s.uppercased())
            .font(.system(size: 11, weight: .bold))
            .tracking(2)
            .foregroundStyle(p.inkFaint)
    }
}

/// An answer, rendered properly: headline card when there is one, then formatted prose.
///
/// The prose is not dumped as one block. Headings, bullets and bold are picked out, because a wall
/// of unbroken text is the difference between an answer that looks considered and one that looks
/// like a log line.
struct AnswerView: View {
    let text: String
    var showHero: Bool = true

    @Environment(\.palette) private var p

    var body: some View {
        VStack(alignment: .leading, spacing: T.md) {
            if showHero, let hero = Hero.detect(in: text) {
                HeroCardView(hero: hero)
            }
            VStack(alignment: .leading, spacing: 10) {
                ForEach(Array(blocks.enumerated()), id: \.offset) { _, block in
                    block.view(p)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .textSelection(.enabled)
    }

    private enum Block {
        case heading(String)
        case bullet(String)
        case numbered(String, String)
        case paragraph(String)

        @ViewBuilder
        func view(_ p: Palette) -> some View {
            switch self {
            case .heading(let s):
                Text(inline(s, p))
                    .font(.system(size: T.body, weight: .semibold))
                    .foregroundStyle(p.ink)
                    .padding(.top, 4)
            case .bullet(let s):
                HStack(alignment: .top, spacing: 10) {
                    Circle().fill(p.accent).frame(width: 5, height: 5).padding(.top, 7)
                    Text(inline(s, p)).font(.system(size: T.body)).foregroundStyle(p.ink)
                }
            case .numbered(let n, let s):
                HStack(alignment: .top, spacing: 10) {
                    Text(n).font(.system(size: T.small, weight: .semibold)).foregroundStyle(p.accent)
                        .frame(minWidth: 16, alignment: .trailing)
                    Text(inline(s, p)).font(.system(size: T.body)).foregroundStyle(p.ink)
                }
            case .paragraph(let s):
                Text(inline(s, p)).font(.system(size: T.body)).foregroundStyle(p.ink)
            }
        }

        /// `**bold**` and `*italic*` become real emphasis rather than visible asterisks — models
        /// emit markdown whether or not you ask them to.
        private func inline(_ s: String, _ p: Palette) -> AttributedString {
            (try? AttributedString(
                markdown: s,
                options: .init(interpretedSyntax: .inlineOnlyPreservingWhitespace)))
                ?? AttributedString(s)
        }
    }

    private var blocks: [Block] {
        text.split(separator: "\n", omittingEmptySubsequences: false).compactMap { raw in
            let line = raw.trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { return nil }

            if let m = line.firstMatch(#"^#{1,6}\s+(.*)$"#) { return .heading(m[1]) }
            if let m = line.firstMatch(#"^[-•*]\s+(.*)$"#) { return .bullet(m[1]) }
            if let m = line.firstMatch(#"^(\d{1,2})[.)]\s+(.*)$"#) { return .numbered(m[1] + ".", m[2]) }
            // A short line that is entirely bold reads as a heading, which is how models write them.
            if let m = line.firstMatch(#"^\*\*(.{2,60})\*\*:?$"#) { return .heading(m[1]) }
            return .paragraph(line)
        }
    }
}

extension StringProtocol {
    /// Regex capture groups by index, with `[0]` the whole match. Returns nil when there is no match,
    /// so call sites read as `if let m = … { m[1] }`.
    func firstMatch(_ pattern: String) -> [String]? {
        let s = String(self)
        guard let re = try? NSRegularExpression(pattern: pattern),
              let m = re.firstMatch(in: s, range: NSRange(s.startIndex..., in: s))
        else { return nil }
        return (0..<m.numberOfRanges).map { i in
            guard let r = Range(m.range(at: i), in: s) else { return "" }
            return String(s[r])
        }
    }
}
